File: std_normals.py
import numpy as np
from numba import jit
import time


@jit(fastmath=True)
def std_normals_numba(n):
    x = np.empty(n)
    for i in range(n):
        x[i] = np.random.randn()
    x = x * x
    total = 0.0
    for i in range(n):
        total += x[i]
    mean = total / n
    var_sum = 0.0
    for i in range(n):
        var_sum += (x[i] - mean) ** 2
    std_dev = np.sqrt(var_sum / (n - 1))
    return mean, std_dev


n = 100_000_000

# Warm up JIT
std_normals_numba(10)

start = time.time()
mean, std_dev = std_normals_numba(n)
elapsed = time.time() - start

print(f"Mean of squared normals:    {mean:12.6f}")
print(f"Std dev of squared normals: {std_dev:12.6f}")
print(f"Elapsed time:  {elapsed:10.4f} seconds")

File: std_normals.f90
program std_normals
    implicit none

    integer, parameter :: n = 100000000
    real(8), allocatable :: x(:)
    real(8) :: mean, variance, std_dev
    real(8) :: u1, u2, pi
    integer :: i
    integer :: t_start, t_end, count_rate

    pi = 4.0d0 * atan(1.0d0)

    allocate(x(n))

    call system_clock(t_start, count_rate)

    ! Generate standard normals via Box-Muller transform
    do i = 1, n, 2
        call random_number(u1)
        call random_number(u2)
        x(i)   = sqrt(-2.0d0 * log(u1)) * cos(2.0d0 * pi * u2)
        if (i + 1 <= n) then
            x(i+1) = sqrt(-2.0d0 * log(u1)) * sin(2.0d0 * pi * u2)
        end if
    end do

    ! Square them
    x = x * x

    ! Compute mean
    mean = sum(x) / dble(n)

    ! Compute standard deviation
    variance = sum((x - mean)**2) / dble(n - 1)
    std_dev = sqrt(variance)

    call system_clock(t_end)

    print '(A, F12.6)', 'Mean of squared normals:   ', mean
    print '(A, F12.6)', 'Std dev of squared normals:', std_dev
    print '(A, F10.4, A)', 'Elapsed time: ', &
        dble(t_end - t_start) / dble(count_rate), ' seconds'

    deallocate(x)

end program std_normals
